yaml/service.go

```diff
@@ -58,21 +60,78 @@ func (s *ServiceSlice) ToPipeline() *pipeline.ContainerSlice {
 // UnmarshalYAML implements the Unmarshaler interface for the ServiceSlice type.
 //
 //nolint:dupl // accepting duplicative code that exists in step.go as well
-func (s *ServiceSlice) UnmarshalYAML(unmarshal func(interface{}) error) error {
+func (s *ServiceSlice) UnmarshalYAML(v *yaml.Node) error {
+	// service slice should be sequence
+	if v.Kind != yaml.SequenceNode {
+		return fmt.Errorf("invalid yaml: expected sequence node for service slice")
+	}
+
 	// service slice we try unmarshalling to
 	serviceSlice := new([]*Service)
 
-	// attempt to unmarshal as a service slice type
-	err := unmarshal(serviceSlice)
-	if err != nil {
-		return err
-	}
-
 	// iterate through each element in the service slice
-	for _, service := range *serviceSlice {
-		// handle nil service to avoid panic
-		if service == nil {
-			return fmt.Errorf("invalid service with nil content found")
+	for _, st := range v.Content {
+		// make local var
+		tmpService := *st
+
+		// services are mapping nodes
+		if tmpService.Kind != yaml.MappingNode {
+			return fmt.Errorf("invalid yaml: expected map node for service")
+		}
+
+		// initialize anchor node -- will be nil if `<<` never used
+		var anchorKey *yaml.Node
+
+		// initialize anchor sets
+		anchorList := new([]*yaml.Node)  // collect multiple anchor references
+		newContent := new([]*yaml.Node)  // new service content
+		anchorSequence := new(yaml.Node) // final type that is appended to service contents
+
+		// iterate through map contents (key, value)
+		for i := 0; i < len(tmpService.Content); i += 2 {
+			key := tmpService.Content[i]
+			value := tmpService.Content[i+1]
```

Comment:
im assuming this cant panic? i couldnt get it to be nil while testing

Reply:
in the same spirit, the length is guaranteed to be an even number? (i think so?) :D